### PR TITLE
JavaFX Client & Controller handling

### PR DESCRIPTION
**Pull request by hendrikebbers**
This merge requests add some convenience classes (like DolphinPlatformApplication) to the JavaFX client API and solves some problems in the creation and destruction of controller proxies and the client context

[Review on Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/13)

### CODE REVIEW

File: java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/AbstractViewBinder.java [hunk 1]
```diff
@@ -15,11 +15,11 @@
  */
 package com.canoo.dolphin.client.javafx;
 
-import com.canoo.dolphin.client.ClientContext;
-import com.canoo.dolphin.client.ControllerActionException;
-import com.canoo.dolphin.client.ControllerProxy;
-import com.canoo.dolphin.client.Param;
-import javafx.beans.property.*;
+import com.canoo.dolphin.client.*;
```
Comment (netopyr) [inline java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/AbstractViewBinder.java]: Another * import

**Comment by aalmiray** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[java-client/src/main/java/com/canoo/dolphin/client/ClientInitializationException.java, line 1 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8Zzpv2-SgILlxgXRLN:-K8Zzpv2-SgILlxgXRLO:-1617560782) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/ClientInitializationException.java#L1)):*
all source code must have ASL2 license headers

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 19 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8ZzycBYY5spIEdM9zr:-K8ZzycBYY5spIEdM9zs:-756711771) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L19)):*
star imports again

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 49 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8_-DOWcx4FLo2Q8f_z:-K8_-DOWcx4FLo2Q8fa-:-23100040) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L49)):*
could be made `final`

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 70 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8_-5BE1Xuv_A4RVW1i:-K8_-5BF0cz-w_cvqHzB:470559537) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L70)):*
must the whole method be marked as `synchronized`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by netopyr** [comment]
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 18 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bplll3oZM7ElQ1JUD:-K8bpllmSCszRts6boWe:2098689027) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L18)):*
Not needed. Application.init() is empty by definition.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 22 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bq2IEElqoxUCdJLey:-K8bq2IEElqoxUCdJLez:1927673606) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L22)):*
Cannot...

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 30 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bqWNpTmEgWhPtmaNt:-K8bqWNpTmEgWhPtmaNu:1326169711) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L30)):*
Is String really the right type for the server endpoint?

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 47 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8brmfXVKCO7qKsML0l:-K8brmfXVKCO7qKsML0m:1616374833) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L47)):*
Should be protected, no?

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8brQSKgEB_sTcxXMzW:-K8brQSKgEB_sTcxXMzX:-1510213251) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L56)):*
As a rule of thumb, we should always use a timeout when waiting for asynchronous functionality to complete.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bspAqxYLUaP7lOiIf:-K8bspAqxYLUaP7lOiIg:901818520) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L56)):*
I wonder if we need to give the user a chance to do something on shutdown, but before the clientContext is disconnected.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 70 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bsCo0XTOXipmhfn2O:-K8bsCo0XTOXipmhfn2P:628437222) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L70)):*
Is this really needed?

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cUUTX14iOau9v_C5k:-K8cUUTYG9c0dt2wufNW:703313457) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L56)):*
Why CopyOnWrite?

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 123 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVD0bMirD7hhdSkoR:-K8cVD0bMirD7hhdSkoS:44153554) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L123)):*
Please add a timeout

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 128 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVKLU1p-casP8EfWt:-K8cVKLU1p-casP8EfWu:-957544808) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L128)):*
I think it would be better to use an Iterator at this point, which allows you to remove the current element.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 137 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVq3gHMSTM6RcDRmL:-K8cVq3gHMSTM6RcDRmM:-1862232546) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L137)):*
I think obtrudeException() is not correct here

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 145 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cW0xd8_Z6MzCbyWwU:-K8cW0xd8_Z6MzCbyWwV:1284817615) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L145)):*
Please use a switch-case statement

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by hendrikebbers** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 30 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bqWNpTmEgWhPtmaNt:-K8dfB9HD7AYbac1fyMD:551658806) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L30)):*
OD uses String here. We could use URL and then call toString() when starting OD. But I don't want to change this now since it's not part of this issue.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8brQSKgEB_sTcxXMzW:-K8dfsSJiCQjvrahDRH3:-473760816) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L56)):*
Sounds like a good idea. I will open an issue for that.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bspAqxYLUaP7lOiIf:-K8dg-S2Hlkd5hx9SB-d:1583046362) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L56)):*
good question. I think that we can simply add this if its needed, right?

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 70 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bsCo0XTOXipmhfn2O:-K8dhR5gOzQ_SDoCNWKD:-508991850) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L70)):*
If I don't do it the Java Process will still run after I closed the Stage. I always have this with JavaFX (when starting an app with IntelliJ)

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cUUTX14iOau9v_C5k:-K8di9sqzvfCvhIRL7w7:-1691106640) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L56)):*
Because I remove elements in an iteration on the list

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 70 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8_-5BE1Xuv_A4RVW1i:-K8dinoNrUOFjsCv2wvf:-1889564541) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L70)):*
Created an issue

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 123 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVD0bMirD7hhdSkoR:-K8diqQCWkNm1nsmzcqI:-274257206) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L123)):*
I created an issue for that.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 128 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVKLU1p-casP8EfWt:-K8dj7_UYrhdHgIAhOPA:838191605) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L128)):*
I don't know what you mean.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 137 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVq3gHMSTM6RcDRmL:-K8djHEu1eLkfRYGFUC9:677194780) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L137)):*
Why?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by hendrikebbers** [comment]
Review status: 4 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 30 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bqWNpTmEgWhPtmaNt:-K8dl9oNMCxV30pfC6NO:-1735752214) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L30)):*
I created an issue at github

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8brQSKgEB_sTcxXMzW:-K8dlETW-V-3ED28DW9n:-499146009) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L56)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by netopyr** [comment]
Reviewed 3 of 5 files at r2.
Review status: 7 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java, line 70 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8bsCo0XTOXipmhfn2O:-K8drj29n2FP2ayqUqbK:-174631401) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client-javafx/src/main/java/com/canoo/dolphin/client/javafx/DolphinPlatformApplication.java#L70)):*
Sounds more like an issue in IntelliJ. Anyway, I guess this does not hurt.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 56 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cUUTX14iOau9v_C5k:-K8drLuaGeQSkRvS__uF:302565451) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L56)):*
Use a listIterator instead

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 128 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVKLU1p-casP8EfWt:-K8dp0amwRMKvbsc30us:-606640839) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L128)):*
If one wants to iterate over a List and remove elements while in the loop, I think it is better to use a listIterator. The iterator has a remove() method that removes the current element. 

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 137 \[r1\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8cVq3gHMSTM6RcDRmL:-K8dpy1i9-n3Ddp6mNyf:-1132330050) ([raw file](https://github.com/canoo/dolphin-platform/blob/e091a65a62658dfd061372ef0ce1e1296056fd91/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L137)):*
Quite frankly, I think obtrudeException() and obtrudeValue() are broken concepts and should not be used ever. If you feel the need to use them, something is probably wrong about your logic. Even the JavaDocs acknowledge their fragility already.

I think instead you should move the result.complete(null)-statement to the very end of the try-block and use completeExceptionally() in the catch-block.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by netopyr** [comment]
Reviewed 1 of 5 files at r2.
Review status: 8 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by netopyr** [comment]
*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 47 \[r2\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8xFY6u0goHmrf8yarw:-K8xFY6u0goHmrf8yarx:-1126826482) ([raw file](https://github.com/canoo/dolphin-platform/blob/3a12cc23964565cf1c0737b2730bcc395fc3b5e9/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L47)):*
Unfortunately implementing this functionality in JavaScript does not seem to be possible, because we rely on an iterable collection of weak references. AFAIK this is not available in JavaScript. We can have either iterable or weak collections, but not both.

We need to reconsider this feature.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by netopyr** [comment]
Review status: 8 of 9 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 47 \[r2\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K8xFY6u0goHmrf8yarw:-K92ACz-kkRGZ4-99laO:-1651606059) ([raw file](https://github.com/canoo/dolphin-platform/blob/3a12cc23964565cf1c0737b2730bcc395fc3b5e9/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L47)):*
I was not able to come up with a good solution. Guess in JavaScript one has to destroy a controller explicitly or they will remain until DP is disconnected. At least for the Polymer client, I can probably automate this on attach and detach events.

---

*[java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java, line 57 \[r2\]](https://reviewable.io:443/reviews/canoo/dolphin-platform/13#-K92JWldaJioaJ_Ml28A:-K92JWldaJioaJ_Ml28B:1958442036) ([raw file](https://github.com/canoo/dolphin-platform/blob/3a12cc23964565cf1c0737b2730bcc395fc3b5e9/java-client/src/main/java/com/canoo/dolphin/client/impl/ClientContextImpl.java#L57)):*
While thinking about a solution for the JavaScript client, I realized that this solution can be improved by automatically calling destroy on any ControllerProxy that is about to be garbage collected. This can be implemented using a ReferenceQueue.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canoo/dolphin-platform/13)*

**Comment by hendrikebbers** [comment]
I think this should be moved to version 0.9 because it's not easy to find a way that will work in Java & JS. By doing so we can discuss this after 0.8 has been released. What do you think?

**Comment by netopyr** [comment]
What we already have will fix a blocker in our application, thus I would prefer if we put the current state in 0.8 and improve on that in 0.9.

**Comment by hendrikebbers** [comment]
By doing so we will have this new auto clean only for Java clients, right?

**Comment by netopyr** [comment]
Well, the auto clean is only working partially IMO, because we do not call destroy() on ControllerProxies, which are about to be GCed.

In JavaScript we will have at least an automated cleanup on disconnecting, which is probably all we will ever get.

**Comment by hendrikebbers** [comment]
In this case the request can be merged, right?

**Comment by hendrikebbers** [comment]
Oh, there are still some open points in the review. Don't finished it because of the Java - JS discussion. Will work on the review.

**Pull request merged**